test_conformance/subgroups/test_subgroup_ballot.cpp

```diff
@@ -31,45 +31,93 @@ template <typename Ty> struct BALLOT
 
     static void gen(Ty *x, Ty *t, cl_int *m, const WorkGroupParams &test_params)
     {
-        // no work here
         int gws = test_params.global_workgroup_size;
         int lws = test_params.local_workgroup_size;
         int sbs = test_params.subgroup_size;
+        int sb_number = (lws + sbs - 1) / sbs;
         int non_uniform_size = gws % lws;
+        int wg_number = gws / lws;
+        wg_number = non_uniform_size ? wg_number + 1 : wg_number;
+        int last_subgroup_size = 0;
+
+        for (int wg_id = 0; wg_id < wg_number; ++wg_id)
+        { // for each work_group
+            if (non_uniform_size && wg_id == wg_number - 1)
+            {
+                set_last_workgroup_params(non_uniform_size, sb_number, sbs, lws,
+                                          last_subgroup_size);
+            }
+            for (int sb_id = 0; sb_id < sb_number; ++sb_id)
+            { // for each subgroup
+                int wg_offset = sb_id * sbs;
+                int current_sbs;
+                if (last_subgroup_size && sb_id == sb_number - 1)
+                {
+                    current_sbs = last_subgroup_size;
+                }
+                else
+                {
+                    current_sbs = wg_offset + sbs > lws ? lws - wg_offset : sbs;
+                }
+
+                for (int wi_id = 0; wi_id < current_sbs; wi_id++)
+                {
+                    cl_uint v;
+                    if (genrand_bool(gMTdata))
+                    {
+                        v = genrand_bool(gMTdata);
+                    }
+                    else if (genrand_bool(gMTdata))
+                    {
+                        v = 1U << ((genrand_int32(gMTdata) % 31) + 1);
+                    }
+                    else
+                    {
+                        v = genrand_int32(gMTdata);
+                    }
```

Comment:
Not sure that the probabilities on this sequence match what's expected, as from what i can tell, the first branch is executed 50% of the time, and the following two branches 25% each.

It might not matter at all, but I felt that perhaps I should mention it.

Reply:
Yeah, that seemed fairly reasonable to me at the time I wrote it. It certainly didn't seem worth doing anything more complicated. The point here is that if we just used random 32-bit integers, they would be true 99.9999999767% of the time. The testing should be improved at some point so that with wide warp widths, there's a reasonable chance of the predicate being the same for all threads... but for now, this is better than what we had.

Reply:
Assuming there's no objection here.